src/theory/strings/theory_strings_rewriter.cpp

```diff
@@ -1599,7 +1676,78 @@ Node TheoryStringsRewriter::rewriteContains( Node node ) {
     Node ret = NodeManager::currentNM()->mkConst(false);
     return returnRewrite(node, ret, "ctn-len-ineq");
   }
-  else if (checkEntailArithEq(len_n1, len_n2))
+  
+  // multi-set reasoning
+  //   For example, contains( str.++( x, "b" ), str.++( "a", x ) ) ---> false
+  //   since the number of a's in the second argument is greater than the number
+  //   of a's in the first argument
+  std::map< Node, unsigned > num_nconst[2];
+  std::map< Node, unsigned > num_const[2];
+  for( unsigned j=0; j<2; j++ ){
+    std::vector< Node >& ncj = j==0 ? nc1 : nc2;
+    for( const Node& cc : ncj ){
+      if( cc.isConst() ){
+        num_const[j][cc]++;
+      }else{
+        num_nconst[j][cc]++;
+      }
+    }
+  }
+  bool ms_success = true;
+  for( std::pair< const Node, unsigned >& nncp : num_nconst[0] )
+  {
+    if( nncp.second>num_nconst[1][nncp.first] ){
+      ms_success = false;
+      break;
+    }
+  }
+  if( ms_success ){
+    // count the number of constant characters in the first argument
+    std::map< Node, unsigned > count_const[2];
+    std::vector< Node > chars;
+    for( unsigned j=0; j<2; j++ ){
+      for( std::pair< const Node, unsigned >& ncp : num_const[j] )
+      {
+        Node cn = ncp.first;
+        Assert(cn.isConst());
+        const std::vector<unsigned>& cvec = cn.getConst<String>().getVec();
+        for( unsigned i=0, size = cvec.size(); i<size; i++ )
+        {
+          // make the character
+          std::vector< unsigned > cc_vec;
```

Comment:
(If we keep the `chars` vector in its current form) Consider moving `cc_vec` outside of this loop and replace the element in it instead of creating a new vector on each iteration?

Reply:
Same as above, I wrote it with the intention to be generalized at some point, where cc_vec would take n characters if we are doing the rewrite based on substrings of length n.

Reply:
Moving `cc_vec` outside of the loops and clearing it might still be a [bit more efficient](https://stackoverflow.com/questions/19189699/clearing-a-vector-or-defining-a-new-vector-which-one-is-faster) because there are going to be less allocations but that's nothing urgent.